wiring/src/spark_wiring_spi.cpp

```diff
@@ -192,59 +256,126 @@ void SPIClass::computeClockDivider(unsigned reference, unsigned targetSpeed, uin
 
 unsigned SPIClass::setClockSpeed(unsigned value, unsigned value_scale)
 {
-    // actual speed is the system clock divided by some scalar
-    unsigned targetSpeed = value*value_scale;
-    hal_spi_info_t info;
-    querySpiInfo(_spi, &info);
-    uint8_t rate;
     unsigned clock;
-    computeClockDivider(info.system_clock, targetSpeed, rate, clock);
-    HAL_SPI_Set_Clock_Divider(_spi, rate);
+
+    // actual speed is the system clock divided by some scalar
+    unsigned targetSpeed = value * value_scale;
+
+    if (!lock())
+    {
+        // Query SPI info
+        hal_spi_info_t info;
+        querySpiInfo(_spi, &info);
+
+        // Calculate clock divider
+        uint8_t divisor = 0;
+        computeClockDivider(info.system_clock, targetSpeed, divisor, clock);
+
+        // Update SPI peripheral
+        HAL_SPI_Set_Clock_Divider(_spi, divisor);
+        unlock();
+    }
+    else
+    {
+        // Unable to acquire lock
+        clock = 0;
+    }
+
     return clock;
 }
 
 byte SPIClass::transfer(byte _data)
 {
-  return HAL_SPI_Send_Receive_Data(_spi, _data);
+    uint16_t result;
+    if (!lock())
+    {
+        result = HAL_SPI_Send_Receive_Data(_spi, _data);
+        unlock();
+    }
+    else
+    {
+        // zero is an error code `HAL_SPI_DMA_Transfer_Status`
+        result = 0;
```

Comment:
I'm accustomed to the explicit statement of return values for each branch, this ensures you have considered each path. Also, the compiler can warn you, if you leave a result branch uninitialized (i.e. unconsidered).

Reply:
I'm afraid I will have to insist on this one. It is an error-prone approach to not explicitly initialize variables with trivial types in almost all the cases.

> Also, the compiler can warn you, if you leave a result branch uninitialized (i.e. unconsidered).
We do not promote warnings to errors in user applications, so it is especially dangerous in wiring.

We also did have a lengthy discussion about the usage of this waterfall if/else pattern both during the PR reviews and at one of the guild meetings and collectively came to a decision that we should not use it as breaks RAII usage and decreases readability. It's almost always better to return/error/throw early, instead of having to jump between if/else code blocks to figure out the execution flow. Inability to use `CHECK()` macros in certain cases also does not prevent from returning/erroring out early.

Reply:
> We do not promote warnings to errors in user applications, so it is especially dangerous in wiring.

I had not considered this ^^^. Without being able to rely on the compiler, I agree this technique is more dangerous than helpful.